### PR TITLE
Torchani lazy model loading

### PR DESCRIPTION
**Pull request by dotsdl**
## Description

Only load/instantiate TorchANI models when one is actually used. Currently, all models get instantiated, even though only one ultimately gets used (and cached) in a `compute` call.

Besides the performance hit, the current implementation also appears to cause frequent issues on networked filesystems due to the many individual file reads required to instantiate the models. For [example](https://github.com/openforcefield/qca-dataset-submission/pull/136#issuecomment-687116908):

```
count : 578

geomeTRIC run_json error:
Traceback (most recent call last):
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/run_json.py", line 225, in geometric_run_json
geometric.optimize.Optimize(coords, M, IC, engine, None, params)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/optimize.py", line 1331, in Optimize
return optimizer.optimizeGeometry()
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/optimize.py", line 1293, in optimizeGeometry
self.calcEnergyForce()
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/optimize.py", line 1002, in calcEnergyForce
spcalc = self.engine.calc(self.X, self.dirname)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/engine.py", line 873, in calc
return self.calc_new(coords, dirname)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/geometric/engine.py", line 865, in calc_new
raise QCEngineAPIEngineError("QCEngineAPI computation did not execute correctly. Message: " + ret["error"]["error_message"])
geometric.errors.QCEngineAPIEngineError: QCEngineAPI computation did not execute correctly. Message: QCEngine Execution Error:
Traceback (most recent call last):
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/qcengine/util.py", line 115, in compute_wrapper
yield metadata
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/qcengine/compute.py", line 91, in compute
output_data = executor.compute(input_data, config)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/qcengine/programs/torchani.py", line 110, in compute
model = self.get_model(method)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/qcengine/programs/torchani.py", line 76, in get_model
"ani1ccx": torchani.models.ANI1ccx(),
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/torchani/models.py", line 312, in ANI1ccx
return BuiltinEnsemble._from_neurochem_resources(info_file, periodic_table_index)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/torchani/models.py", line 238, in _from_neurochem_resources
ensemble_prefix, ensemble_size)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/torchani/neurochem/__init__.py", line 271, in load_model_ensemble
models.append(load_model(species, network_dir))
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/torchani/neurochem/__init__.py", line 253, in load_model
models[i] = load_atomic_network(filename)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/torchani/neurochem/__init__.py", line 232, in load_atomic_network
load_param_file(layer, in_size, out_size, wfn, bfn)
File "/data/homezvol0/tgokey/.local/miniconda3/envs/qcf/lib/python3.7/site-packages/torchani/neurochem/__init__.py", line 203, in load_param_file
w = struct.unpack('{}f'.format(wsize), fw.read())
struct.error: unpack requires a buffer of 81920 bytes
```

## Changelog description

`TorchANIHarness` now lazily loads models as requested for compute.

## Status

- [ ] Code base linted
- [ ] Ready to go

### CODE REVIEW

**Comment by dotsdl** [comment]
@trevorgokey I think this may substantially reduce the frequency of the deserialization errors we are seeing for our `ani2x` dataset. Not certain this is *the* cause, but it seems likely to me based on the information we have in front of us.

**Comment by trevorgokey** [comment]
Always wondered why the tracebacks were going through `ANI1ccx`; this PR rightfully skips this code path.

I have manually applied this PR, and (so far) I am no longer seeing the above error. Nice catch!

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -13,6 +13,19 @@ Changelog
...
+New Features
+++++++++++++
+
+Enhancements
+++++++++++++
+- (:pr:`239`) OpenMM - extracts CMILES from ``AtomicInput`` if possible for consistent atom ordering
+
+Bug Fixes
++++++++++
+
```
Comment (loriab) [inline docs/source/changelog.rst:28]: feel free to start the v0.17 changelog, but this looks like a forgotten start.

File: qcengine/programs/torchani.py [hunk 2]
```diff
@@ -72,15 +72,15 @@ def ensemble_forward(self, species_input):
...
+            "ani1ccx": torchani.models.ANI1ccx,
         }
 
         if parse_version(self.get_version()) >= parse_version("2.0"):
-            ani_models["ani2x"] = torchani.models.ANI2x()
+            ani_models["ani2x"] = torchani.models.ANI2x
 
         try:
-            self._CACHE[name] = ani_models[name]
+            self._CACHE[name] = ani_models[name]()
```
Comment (loriab) [inline qcengine/programs/torchani.py:83]: good catch!

File: docs/source/changelog.rst (hunk 1 above)
Comment (dotsdl) [reply docs/source/changelog.rst]: Yeah sorry, this came from a polluted master branch on my local repo. Cleared it out.

**Comment by dotsdl** [comment]
Changelog updated! Thanks @loriab!

**Review by loriab — APPROVED** [review]
merge at will

**Pull request merged**